// easycodef-node/dist/accesstoken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessToken = void 0;
var constant_1 = require("./constant");
/**
 *  서비스 별 토큰 관리 클래스
 */
var AccessToken = /** @class */ (function () {
    function AccessToken() {
        this.sandbox = '';
        this.demo = '';
        this.api = '';
    }
    AccessToken.prototype.setToken = function (serviceType, accessToken) {
        switch (serviceType) {
            case constant_1.SERVICE_TYPE_API:
                this.api = accessToken;
                break;
            case constant_1.SERVICE_TYPE_DEMO:
                this.demo = accessToken;
                break;
            default:
                this.sandbox = accessToken;
        }
    };
    AccessToken.prototype.getToken = function (serviceType) {
        switch (serviceType) {
            case constant_1.SERVICE_TYPE_API:
                return this.api;
            case constant_1.SERVICE_TYPE_DEMO:
                return this.demo;
            default:
                return this.sandbox;
        }
    };
    return AccessToken;
}());
exports.AccessToken = AccessToken;
