// easycodef-node/dist/constant.js
"use strict";
/************************************************************************************************************
 *  코드에프 게정 정보
 ************************************************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.DELETE_ACCOUNT = exports.UPDATE_ACCOUNT = exports.ADD_ACCOUNT = exports.CREATE_ACCOUNT = exports.GET_ACCOUNT_LIST = exports.GET_CID_LIST = exports.SERVICE_TYPE_SANDBOX = exports.SERVICE_TYPE_DEMO = exports.SERVICE_TYPE_API = exports.ACCESS_DENIED = exports.INVALID_TOKEN = exports.DATA = exports.EXTRA_MESSAGE = exports.MESSAGE = exports.CODE = exports.RESULT = exports.SANDBOX_CLIENT_SECRET = exports.SANDBOX_CLIENT_ID = exports.SANDBOX_DOMAIN = exports.DEMO_DOMAIN = exports.API_DOMAIN = exports.REPEAT_COUNT = exports.GET_TOKEN = exports.OAUTH_DOMAIN = void 0;
/**	OAUTH 서버 도메인	*/
exports.OAUTH_DOMAIN = 'https://oauth.codef.io';
/**	OAUTH 엑세스 토큰 발급 URL PATH	*/
exports.GET_TOKEN = '/oauth/token';
/**   토큰발급 재시도 3회  **/
exports.REPEAT_COUNT = 3;
/**    정식 서버 도메인    */
exports.API_DOMAIN = 'https://api.codef.io';
/**	데모 서버 도메인	*/
exports.DEMO_DOMAIN = 'https://development.codef.io';
/**	샌드박스 서버 도메인	*/
exports.SANDBOX_DOMAIN = 'https://sandbox.codef.io';
exports.SANDBOX_CLIENT_ID = 'ef27cfaa-10c1-4470-adac-60ba476273f9';
exports.SANDBOX_CLIENT_SECRET = '83160c33-9045-4915-86d8-809473cdf5c3';
/************************************************************************************************************
 *  코드에프 응답 코드
 *************************************************************************************************************/
/** 응답부 수행 결과 키워드	*/
exports.RESULT = 'result';
/** 응답부 수행 결과 메시지 코드 키워드	*/
exports.CODE = 'code';
/** 응답부 수행 결과 메시지 키워드	*/
exports.MESSAGE = 'message';
/** 응답부 수행 결과 추가 메시지 키워드	*/
exports.EXTRA_MESSAGE = 'extraMessage';
/**	응답부 수행 결과 데이터 키워드	*/
exports.DATA = 'data';
/**	엑세스 토큰 거절 사유1	*/
exports.INVALID_TOKEN = 'invalid_token';
/**	엑세스 토큰 거절 사유2	*/
exports.ACCESS_DENIED = 'access_denied';
exports.SERVICE_TYPE_API = 0; // 정식
exports.SERVICE_TYPE_DEMO = 1; // 데모
exports.SERVICE_TYPE_SANDBOX = 2; // 샌드박스
/************************************************************************************************************
 *  코드에프 상품 URL 정보
 *************************************************************************************************************/
// /**	요청 상품 주소	*/
// PRODUCT_URL: "",
/**	커넥티드 아이디 목록 조회 URL	*/
exports.GET_CID_LIST = '/v1/account/connectedId-list';
/**	계정 목록 조회 URL	*/
exports.GET_ACCOUNT_LIST = '/v1/account/list';
/**	계정 등록 URL	*/
exports.CREATE_ACCOUNT = '/v1/account/create';
/**	계정 추가 URL	*/
exports.ADD_ACCOUNT = '/v1/account/add';
/**	계정 수정 URL	*/
exports.UPDATE_ACCOUNT = '/v1/account/update';
/**	계정 삭제 URL	*/
exports.DELETE_ACCOUNT = '/v1/account/delete';
exports.default = {
    OAUTH_DOMAIN: exports.OAUTH_DOMAIN,
    GET_TOKEN: exports.GET_TOKEN,
    REPEAT_COUNT: exports.REPEAT_COUNT,
    API_DOMAIN: exports.API_DOMAIN,
    DEMO_DOMAIN: exports.DEMO_DOMAIN,
    SANDBOX_DOMAIN: exports.SANDBOX_DOMAIN,
    SANDBOX_CLIENT_ID: exports.SANDBOX_CLIENT_ID,
    SANDBOX_CLIENT_SECRET: exports.SANDBOX_CLIENT_SECRET,
    RESULT: exports.RESULT,
    CODE: exports.CODE,
    MESSAGE: exports.MESSAGE,
    EXTRA_MESSAGE: exports.EXTRA_MESSAGE,
    DATA: exports.DATA,
    INVALID_TOKEN: exports.INVALID_TOKEN,
    ACCESS_DENIED: exports.ACCESS_DENIED,
    SERVICE_TYPE_API: exports.SERVICE_TYPE_API,
    SERVICE_TYPE_DEMO: exports.SERVICE_TYPE_DEMO,
    SERVICE_TYPE_SANDBOX: exports.SERVICE_TYPE_SANDBOX,
    GET_CID_LIST: exports.GET_CID_LIST,
    GET_ACCOUNT_LIST: exports.GET_ACCOUNT_LIST,
    CREATE_ACCOUNT: exports.CREATE_ACCOUNT,
    ADD_ACCOUNT: exports.ADD_ACCOUNT,
    UPDATE_ACCOUNT: exports.UPDATE_ACCOUNT,
    DELETE_ACCOUNT: exports.DELETE_ACCOUNT,
};
